src/main/java/com/epam/dao/TraineeDao.java

```diff
@@ -1,37 +1,97 @@
 package com.epam.dao;
 
 import com.epam.model.Trainee;
+import com.epam.model.Trainer;
+import com.epam.model.Training;
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.PersistenceContext;
+import jakarta.persistence.Query;
+import jakarta.persistence.TypedQuery;
+import jakarta.persistence.criteria.CriteriaBuilder;
+import jakarta.persistence.criteria.CriteriaQuery;
+import jakarta.persistence.criteria.Root;
+import jakarta.transaction.Transactional;
 import org.springframework.stereotype.Repository;
 
-import java.util.Map;
+import java.util.List;
+import java.util.Optional;
 
 @Repository
-public class TraineeDao extends GenericDao<Trainee> {
-    public static final String TRAINEES = "Trainees";
+@Transactional
+public class TraineeDao {
+    @PersistenceContext
+    EntityManager entityManager;
 
-    @Override
-    public void save(String namespace, Integer id, Trainee entity) {
-        super.save(TRAINEES, id, entity);
+    public Trainee save(Trainee trainee) {
+        entityManager.persist(trainee);
+        return trainee;
     }
 
-    @Override
-    public Trainee findById(String namespace, Integer id) {
-        return super.findById(TRAINEES, id);
+    public Optional<Trainee> update(Trainee updatedEntity) {
+        Trainee mergedTrainee = entityManager.merge(updatedEntity);
+        return Optional.ofNullable(mergedTrainee);
     }
 
-    @Override
-    public Map<Integer, Trainee> findAll(String namespace) {
-        return super.findAll(TRAINEES);
+    public Optional<Trainee> findById(Integer id) {
+        Trainee trainee = entityManager.find(Trainee.class, id);
+        return Optional.ofNullable(trainee);
     }
 
-    @Override
-    public void update(String namespace, Integer id, Trainee updatedEntity) {
-        super.update(TRAINEES, id, updatedEntity);
+    public Optional<Trainee> findByUsername(String username) {
+        TypedQuery<Trainee> query = entityManager.createQuery("""
+                SELECT tr FROM Trainee tr
+                INNER JOIN User u ON u.id = tr.user.id AND u.username LIKE CONCAT('%', :username, '%')
```

Comment:
Will this query return one Trainee, if you have to usernames: 
1) Andy 
2) Andy1993
?